run.go

```diff
@@ -41,18 +51,42 @@ func (a *Autopilot) Stop() <-chan struct{} {
 	defer a.runLock.Unlock()
 
 	// Nothing to do
-	if !a.running {
+	if a.execution == nil || a.execution.status == notRunning {
 		done := make(chan struct{})
 		close(done)
 		return done
 	}
 
-	a.shutdown()
-	return a.done
+	a.execution.shutdown()
+	a.execution.status = shuttingDown
+	return a.execution.done
 }
 
-func (a *Autopilot) run(ctx context.Context) {
+func (a *Autopilot) endRun(exec *execInfo) {
+	// need to gain the lock because if this was the active execution
+	// then these values may be read while they are updated.
+	a.runLock.Lock()
+	defer a.runLock.Unlock()
+
+	exec.shutdown = nil
```

Comment:
Can `exec` ever be nil when you get here?

Reply:
Nope, run is always executed with the `exec` and that is the only place `endRun` should be used. I could have made this an inline func and closed over the exec there instead. Regardless all paths that lead to here will have exec be non-nil